src/types.rs

```diff
@@ -1,51 +1,160 @@
 //! Concrete node and edge types used in the registry.
 
+extern crate num_traits;
+#[cfg(feature = "quickcheck")]
+extern crate quickcheck;
+
+use num_traits::Zero;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::Add;
+
+#[cfg(feature = "quickcheck")]
+use quickcheck::{Arbitrary, Gen};
+
 /// The type of a node.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum NodeType {
     /// A user, eg. contributor, project member etc.
     User,
     /// A project with users as members and contributors.
     Project,
 }
 
+#[cfg(feature = "quickcheck")]
+impl Arbitrary for NodeType {
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        if g.next_u32() % 2 == 0 {
+            Self::User
+        } else {
+            Self::Project
+        }
+    }
+}
+
 /// Node data.
-type NodeData = NodeType;
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct NodeData<W> {
+    /// The type for this node.
+    pub node_type: NodeType,
+    /// The total contributions by this user, to *all* projects, if any.
+    pub total_contributions: Option<u32>,
+    pub rank: NodeRank<W>,
+}
+
+#[cfg(feature = "quickcheck")]
+impl<W> Arbitrary for NodeData<W>
+where
+    W: Arbitrary,
+{
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        NodeData {
+            node_type: Arbitrary::arbitrary(g),
+            total_contributions: Arbitrary::arbitrary(g),
+            rank: Arbitrary::arbitrary(g),
+        }
+    }
+}
 
 /// The type of an edge.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum EdgeType {
-    /// Contribution from a user to a project. This edge type is
-    /// used bi-directionally, between project and user.
-    Contribution,
-    /// Membership of a user in a project, project <-> user.
-    ProjectMembership,
-    /// One-way dependency between two projects, project -> project.
-    ProjectDependency,
+    /// Contribution from a project to a user. Corresponds to `contrib` from the paper.
```

Comment:
```suggestion
    /// Contribution from a user to a project. Corresponds to `contrib` from the paper.
```

Reply:
Let me be proactive and change this to the proposed enumeration -- we can always revert in case 😉

Reply:
Yeah I think this is the way to go :+1: